src/representation_theory/basic.lean

```diff
@@ -0,0 +1,116 @@
+/-
+Copyright (c) 2022 Antoine Labelle. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Labelle
+-/
+import algebra.module.basic
+import data.equiv.module
+import algebra.module.linear_map
+import algebra.monoid_algebra.basic
+import linear_algebra.trace
+
+/-!
+# Monoid representations
+
+This file introduces monoid representations and their characters and proves basic lemmas about them,
+including equivalences between different definitions of representations.
+
+## Main definitions
+
+  * `as_module`
+  * `rep_space`
+  * `character`
+
+## Implementation notes
+
+A representation of a monoid `G` over a commutative semiring `k` is implemented as a `k`-module `V`
+together with a `distrib_mul_action G V` instance and a `smul_comm_class G k V` instance.
+
+Alternatively, one can use a monoid homomorphism `G →* (V →ₗ[k] V)`. The definitions `as_monoid_hom`
+and `rep_space` allow to go back and forth between these two definitions.
+-/
+
+open monoid_algebra
+
+namespace representation
+
+section
+variables (k G V : Type*) [comm_semiring k] [monoid G] [add_comm_monoid V]
+variables [module k V] [distrib_mul_action G V] [smul_comm_class G k V]
+
+/--
+A k-linear representation of G on V can be thought of as
+an algebra map from `monoid_algebra k G` into the k-linear endomorphisms of V.
+-/
+noncomputable def as_algebra_hom : monoid_algebra k G →ₐ[k] (module.End k V) :=
+  (lift k G _) (distrib_mul_action.to_module_End k V)
```

Comment:
This resembles these lines from @101damnations' #11574:

https://github.com/leanprover-community/mathlib/blob/3ae2284c97c345de20f755f3b1f66c1fbd479a19/src/group_theory/group_algebra.lean#L146-L147

Note that the version there works for non-commutative `k`, although maybe no one cares about that case.

Reply:
I'm not sure the case where `k` is non-commutative is particularly interesting since `k`-algebras are only defined when `k` is commutative.